# Introduction to Python for Data Science

**Session Overview**

1. Introduction to Python programming  
2. Data types, variables, and basic operations  
3. Control flow (if/else)  
4. Loops (for, while)

Each section follows an **I Do/Show (My Demo), You Do (Student Exercise), We Do (Class Practice)** pattern depending on the complexity.  

At the end, there are practice exercises to work on independently(This is homework but you do not need to submit, in fact, the challenge will be to practice as assess yourself).

---

## 1. Why Python for Data Science?


  Python is a high‐level, interpreted language with extensive libraries (NumPy, pandas, Matplotlib) that make it ideal for data analysis, visualization, and machine learning.
  
  Python powers many data‐driven projects: from analyzing rainfall patterns in Kisumu, to tracking maize yield in Uasin Gishu, to visualizing Nairobi traffic data.

- **Reasons:**  
  - Readable syntax → faster learning curve  
  - Large ecosystem (pandas, scikit‐learn, etc.)  and community.
  - Used widely in industry & academia  
  - Cross‐platform: works on Windows, macOS, Linux (including Ubuntu on WSL)

---

**House keeping codes**

In [ ]:
# Printing and commenting
# -------------------------------------
# Comments are created by putthing a hash character(#)
print("Everlyn is teaching")

Everlyn is teaching


In [ ]:
# Input
# -------------------------------------
# input() function allows for user input as shown
#
g = int(input("Enter a number: "))

Enter a number: 3


In [ ]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


## 2. Data Types, Variables, and Basic Operations

### 2.1. Python Data Types

Python has several built‐in data types. The most common ones you’ll use in data science are:

1. **Numeric Types**  
   - `int` (integer), e.g., `5`, `2025`  
   - `float` (decimal), e.g., `3.14`, `0.001`  

2. **Text Type**  
   - `str` (string), e.g., `"Nairobi"`, `'Hello, Kenya!'`  

3. **Boolean**  
   - `bool`, either `True` or `False``  

4. **Composite data types: Sequence Types**  
   - `list`, e.g., `[10, 20, 30]`, `['Nairobi', 'Mombasa']`  
   - `tuple`, e.g., `(1, 2, 3)`, `('Kisumu', 'Eldoret')`

5. **Composite data types: Mapping Type**  
   - `dict` (dictionary), e.g., `{'county': 'Nairobi', 'population': 4,400,000}

Often in data analysis, you’ll store numerical arrays (lists), text labels (strings), or key‐value pairs (dictionaries).  

---

In [ ]:
price=['Risper']

### 2.2. Variables

- A **variable** is a name that refers to a value and stores it for easy retrieval.  
- You assign with `=`.  
- Variable names should be descriptive, use lowercase and underscores (e.g., `maize_yield`, `rainfall_mm`).  


In [ ]:
price.append('Manase')

In [ ]:
price

['Risper', 'Kathryn']

In [ ]:
name = 'Risper'
Name = 'M.K'

In [ ]:
Me_Kamilimu='Ds'

In [ ]:
# I Do: Defining variables and checking types
# -------------------------------------------

# 1. Numeric variables
population_nairobi = 4400000      # int: population of Nairobi (approx.)
average_rainfall_kisumu = 127.5   # float: avg rainfall in Kisumu (mm/year)

In [ ]:
# Display variables and their types
print("Type of population_nairobi:", type(population_nairobi))
print("Type of average_rainfall_kisumu:", type(average_rainfall_kisumu))

Type of population_nairobi: <class 'int'>
Type of average_rainfall_kisumu: <class 'float'>


In [ ]:
# 2. String variable
city_name = "Nairobi"

In [ ]:
# 3. List and tuple
counties = ["Nairobi", "Mombasa", "Kisumu"]
port_cities = ("Mombasa", "Malindi", "Lamu")  # tuple

In [ ]:
# 4. Dictionary
county_population = {
    "Nairobi": 4400000,
    "Mombasa": 1200000,
    "Kisumu": 600000
}

In [ ]:
# 5. Boolean
is_urban = True

**You Do: Defining variables and checking types**

In [ ]:
age = 20
print(type(age))

<class 'int'>


In [ ]:
price

['Risper', 'Kathryn', 'Manase']

### 2.2. Basic Operations

- Numeric operations: +, -, * , /, ** (power), % (modulo), // (floor division)
- String concatenation: using + or f-strings
- List operations: indexing, slicing, concatenation

In [ ]:
no_talking = ['Abigael','Faith']

In [ ]:
people_in_class = price + no_talking

In [ ]:
print(people_in_class)

['Risper', 'Kathryn', 'Manase', 'Abigael', 'Faith']


In [ ]:
full_name = f"{price[0]} Audi"
full_name

'Risper Audi'

In [ ]:
# I Do: Demonstrate numeric operations
# -------------------------------------

a = 10
b = 3

sum_ab = a + b            # 13
diff_ab = a - b           # 7
prod_ab = a * b           # 30
div_ab = a / b            # 3.333...
int_div_ab = a // b       # 3
mod_ab = a % b            # 1
power_ab = a ** b         # 10^3 = 1000

print("Sum:", sum_ab)
print("Difference:", diff_ab)
print("Product:", prod_ab)
print("Division:", div_ab)
print("Integer Division:", int_div_ab)
print("Modulo:", mod_ab)
print("Power:", power_ab)

Sum: 13
Difference: 7
Product: 30
Division: 3.3333333333333335
Integer Division: 3
Modulo: 1
Power: 1000


In [ ]:
first_name = "Okiya"
last_name = "Omtatah"
full_name = first_name + " " + last_name
print("Full name (using +):", full_name)

# Using f-strings (Python 3.6+)
greeting = f"Hello, {first_name} {last_name}!"
print("Greeting (using f-string):", greeting)

Full name (using +): Okiya Omtatah
Greeting (using f-string): Hello, Okiya Omtatah!


In [ ]:
# I Do: List indexing, slicing, concatenation
# --------------------------------------------

rainfall_months = [100, 80, 60, 20, 10, 0, 30, 60, 80, 100, 120, 140]  # mm for each month
print("Rainfall in January:", rainfall_months[0])
print("Rainfall in last quarter (Oct-Dec):", rainfall_months[9:12])

# Concatenate two lists
counties_east = ["Meru", "Embu", "Kitui"]
counties_west = ["Kakamega", "Bungoma", "Busia"]
all_counties = counties_east + counties_west
print("All counties (E/W):", all_counties)

Rainfall in January: 100
Rainfall in last quarter (Oct-Dec): [100, 120, 140]
All counties (E/W): ['Meru', 'Embu', 'Kitui', 'Kakamega', 'Bungoma', 'Busia']


**You Do: Student Exercises**

- Create two variables: area of Kenya and population in Kenya
- Compute and print: `population density`

In [ ]:
price.append('Edwin')

In [ ]:
print(no_talking)
no_talking.pop()
print(no_talking)

['Abigael', 'Faith', 'Edwin']
['Abigael', 'Faith']


In [ ]:
area = 580367
population = 55000000

density = population / area
print("Population density in Kenya is", round(density, 2), "people per square kilometer.")

Population density in Kenya is 94.77 people per square kilometer.


In [ ]:
type(population_in_kenya)

tuple

**We do:**



In [ ]:
# We have Ksh 1000, which you invest with a 10% return each year.
# After one year, it's 1000 × 1.1 = 1100 shillings, and after two years
# it's 1000 × 1.1 × 1.1 = 121.
# Now, write code below to calculate how much money
# you would end up with after 7 years.
# How much is your Ksh. 1000 worth after 15 years?
# ----------

principal = 1000
rate = 1.1

after_1 = principal * rate
print(after_1)

after_7 = principal * (rate**7)
print(after_7)

1100.0
1948.7171000000012


In [ ]:
principal = 1000
rate = 0.1

after_1 = principal + (principal * rate * 1)
print(after_1)

after_7 = principal + (principal * rate * 7)
print(after_7)

1100.0
1700.0


## Control Flow (if / elif / else)

- Control flow allows your code to make decisions.

- Basic form:

```
# Structure of if
if <some condition>:
  <specific commands>

# Structure of if/else
if <some condition>:
  <specific commands>
else:
  <other specific commands>

# Structure of if/elif/else
if <some condition>:
  <specific commands>
elif <some other condition>:
  <other specific commands>
elif <some other condition>:
  <other specific commands>
else:
  <other specific commands>
```

- Conditions evaluate to True or False.
- Common comparisons: ==, !=, <, >, <=, >=.
- Combine conditions with and, or, not

In [ ]:
price

['Risper', 'Kathryn', 'Manase', 'Edwin']

In [ ]:
# Example: Classify rainfall level for a given county
county_name = "Kitale"
rainfall_mm = 900  # average annual rainfall

if rainfall_mm > 1000:
    level = "Very High Rainfall"
elif rainfall_mm >= 500:
    level = "Moderate Rainfall"
else:
    level = "Low Rainfall"

print(f"{county_name} has {level} ({rainfall_mm} mm/year).")

Kitale has Moderate Rainfall (900 mm/year).


**You do:**

Check if a county is densely populated.

```
county = "Nairobi"
area = 696  # km²
population = 4400000

density = population / area  # persons per km²
```

In [ ]:
country = "Nairobi"
area = 696  # km²
population = 4400000

density = population / area  # persons per km²


if (density >= 1000):
  print(f"{country} is densely populated")
else:
  print(f"{country} is not densely populated")

Nairobi is densely populated


In [ ]:
county = "Nairobi"
area = 696  # km²
population = 4400000

density = population / area  # persons per km²

if density > 3000:
  print(f"{county}is highly dense.")
else:
  print(f"{county}is not highly dense.")

Nairobiis highly dense.


**We do:**

```
county = "Garissa"
avg_temp_c = 34.0
```

Check is area is hot

In [ ]:
# Write a program that prompts for two numbers, a and b. If the b is not a 0, print the result a/b;
#Otherwise, print division by zero is not allowed.
a=int(input("Enter the first number:"))
b=int(input("Enter the second number:"))
if b !=0:
  print(a/b)
else:
  print("Division by zero isn not allowed")

Enter the first number:24
Enter the second number:5
4.8


In [ ]:
price.append('Eric')

In [ ]:
list(range(0,11,2))

[0, 2, 4, 6, 8, 10]

In [ ]:
price

['Risper', 'Kathryn', 'Manase', 'Edwin', 'Faith', 'Risper', 'Eric']

## Loops (for / while)

- Loops allow you to repeat code blocks.

- `for` loops iterate over sequences (lists, tuples, strings, ranges).

```
# Structure of for loops
for <loop variable> in <iterable>:
  <a command to repeat>
  <maybe more commands to repeat>
```

- `while` loops repeat while a condition remains `True`.

```
# Structure of while loops
while <some condition> :
    <some commands to repeat>
```

- Common usage in DS: filtering data.

In [ ]:
# I Do: for Loop over a List
# --------------------------------------------
for i in price:
  print(i)

Risper
Kathryn
Manase
Edwin
Faith
Risper
Eric


In [ ]:
# I Do: for Loop with range()
# --------------------------------------------
for i in range(6):
    print(i)

0
1
2
3
4
5


In [ ]:
# I Do: while Loop Example, budgeting example
# --------------------------------------------
budget = True
mooney = 10000

while budget:
  item = int(input("Price of item:"))
  print('Buy item')
  mooney -= item
  if mooney<=0:
    budget=False # Exit

Price of item:200
Buy item
Price of item:2000
Buy item
Price of item:5000
Buy item
Price of item:3000
Buy item


**You do/We do:**

`for` loop:

- Create a list of bank balances
- Write a for loop to compute:  
    - Add salary to each balance
    - print out balance after salary increment

`while` loop:
- Simulate water usage and print tank is empty when water is used up.
  - You need total water and usage variable

In [ ]:
for:
while: Manase

In [ ]:
# List of current bank balances
bank_balances = [1200, 2500, 4300, 800, 1500]

# Define a fixed salary to be added
salary = 1000

# Add salary to each balance and print the result
for i in range(len(bank_balances)):
    bank_balances[i] += salary
    print(f"Account {i+1} new balance: {bank_balances[i]}")

Account 1 new balance: 2200
Account 2 new balance: 3500
Account 3 new balance: 5300
Account 4 new balance: 1800
Account 5 new balance: 2500


In [ ]:
balances = [12000,20000,30000,50000]
salary =  40000

for i in range(4):
  balances[i] += salary
  print(f"New {i + 1} balance is: {balances[i]}")

New 1 balance is: 52000
New 2 balance is: 60000
New 3 balance is: 70000
New 4 balance is: 90000


In [ ]:
bank_balances=[10000,8700,12000,33000,3000]
salary=54000
new_balances=[]

for balance in bank_balances:
    new_balance=balance+salary
    new_balances.append(new_balance)
print(new_balances)


[64000, 62700, 66000, 87000, 57000]


In [ ]:
bank_balances = [1200, 850, 940, 1500, 670]

salary = 500

for balance in bank_balances:
    updated_balance = balance + salary
    print(f"Updated balance after salary increment: ${updated_balance}")

Updated balance after salary increment: $1700
Updated balance after salary increment: $1350
Updated balance after salary increment: $1440
Updated balance after salary increment: $2000
Updated balance after salary increment: $1170


In [ ]:
bank_balances = [1500, 2500, 3200, 1800, 4000]

salary = 1000
print("Bank balances before salary increment:", bank_balances)

for i in range(len(bank_balances)):
  bank_balances[i] += salary
  print(f"Balance after salary increment for account {i+1}: {bank_balances[i]}")

print("Bank balances after salary increment:", bank_balances)

Bank balances before salary increment: [1500, 2500, 3200, 1800, 4000]
Balance after salary increment for account 1: 2500
Balance after salary increment for account 2: 3500
Balance after salary increment for account 3: 4200
Balance after salary increment for account 4: 2800
Balance after salary increment for account 5: 5000
Bank balances after salary increment: [2500, 3500, 4200, 2800, 5000]


In [ ]:
total_water = 100

while total_water > 0:
    usage = int(input("Enter the amount of water used (liters): "))
    total_water -= usage

    print(f"Water remaining: {total_water}L")

    if total_water <= 0:
        print("Tank is empty!")

Enter the amount of water used (liters): 50
Water remaining: 50L
Enter the amount of water used (liters): 50
Water remaining: 0L
Tank is empty!


In [ ]:
total_water = 1000 #in litres
water_usage = 50 #per cycle
while total_water > 0:
    total_water -= water_usage
    print(f"Total water left: {total_water} litres")
print("tank is empty")

Total water left: 950 litres
Total water left: 900 litres
Total water left: 850 litres
Total water left: 800 litres
Total water left: 750 litres
Total water left: 700 litres
Total water left: 650 litres
Total water left: 600 litres
Total water left: 550 litres
Total water left: 500 litres
Total water left: 450 litres
Total water left: 400 litres
Total water left: 350 litres
Total water left: 300 litres
Total water left: 250 litres
Total water left: 200 litres
Total water left: 150 litres
Total water left: 100 litres
Total water left: 50 litres
Total water left: 0 litres
tank is empty


In [ ]:
water = 500
use = 100
while water > 0:
    water -= use
    print(water)
print("Empty")

400
300
200
100
0
Empty


In [ ]:
total_water = 10000

while total_water > 0:
    print(f"Remaining water: {total_water} liters")
    water_usage = int(input("Enter water usage (liters): "))
    total_water -= water_usage


print("Tank is empty!")

Remaining water: 10000 liters
Enter water usage (liters): 8000
Remaining water: 2000 liters
Enter water usage (liters): 7000
Tank is empty!


# Two Truths and a Lie: Python Edition

1. “In Python, the expression len([‘Nairobi’, ‘Mombasa’, ‘Kisumu’]) returns 3.”
2. “The variable assignment rainfall = ‘150 mm’ makes rainfall a string, not a number.”
3. “In Python, you can compute average rainfall by writing avg = sum([180, 170, 160]) / 2.”

1. “You can store county‐yield pairs in a dictionary like {'Uasin Gishu': 35.2, 'Nyeri': 32.5}.”
2. “The Python expression if population_kenya = 53771300: is valid syntax to check equality.”
3. “After running counties.append('Kwale') on counties = ['Nairobi', 'Mombasa'], the list becomes ['Nairobi', 'Mombasa', 'Kwale'].”

# Session Recap & Next Steps

Recap:  
- Basic Python data types (int, float, str, list, dict, bool)
- Variables & naming conventions
- Basic arithmetic & string operations
- Control flow (if / elif / else) for decision‐making
- Loops (for, while) for repetition

# Independent practice

## Exercise 1: Simulating Maize Sales

- Suppose a farmer in Kitale sells maize at price_per_kg = 50 Ksh. They have daily harvest amounts for a 7‐day period:

`daily_harvest = [100, 120, 80, 90, 110, 95, 105]`

- Use a while loop to simulate selling day by day. The farmer has a goal of earning at least 5000 Ksh.

  

1.   Keep track of total_earnings = 0 and the day count.
2.   For each day, update total_earnings += daily_harvest[day] * price_per_kg.
3.   Stop once total_earnings ≥ 5000 or all days are processed.  
4.   Print:
    - The day on which the farmer reached or exceeded 5000 Ksh (if at all).
    - Total earnings after 7 days (or when target was met).

In [ ]:
price

['Risper', 'Kathryn', 'Manase', 'Edwin', 'Faith', 'Risper', 'Eric']

## Exercise 2: Tutorial review

Go throught this old notebook and ask questions on slack: https://colab.research.google.com/github/deep-learning-indaba/indaba-pracs-2019/blob/master/Python_Intro.ipynb

## Exercise 3: Simpler challenges

In [ ]:
# Challenge 1
# Many people around the world think about their height in feet and inches
# even here in Africa where many countries use the metric system.
# Write a program reads the height in metres then converts it to
# feet and inches. The program should display the equivalent
# height in feet and height.

In [ ]:
# Challenge 2
# Given a certain column of data that has radius, a data scientist would like
# to compute the volume the respective cylinder and save the result to a new column.
# Thus, write a program that reads the radius of cylinder, along with its height
# from the user to calculate the volume in the console.

In [ ]:
# Challenge 3
# While paying for her parking fee of Ksh. 50 , Jane inserts a Note into
# the paying machine and gets back her change.
# Write a program that reads as input the note greater than Ksh. 200
# and returns change in notes of 100 and 50.
# Use the knowledge of variables to printing, comments, operators
# and input to print the no of denominations given